src/encoder.rs

```diff
@@ -0,0 +1,237 @@
+    fn encode(&self, &[MetricFamily], &mut Write) -> Result<usize>;
+
+    /// `format_type` returns target format.
+    fn format_type(&self) -> &str;
+}
+
+pub type Format = &'static str;
+
+pub const TEXT_FORMAT: Format = "text/plain; version=0.0.4";
+
+/// Implementation of a `Encoder` that converts a `MetricFamily` proto message
+/// into text format.
+#[derive(Debug, Default)]
+pub struct TextEncoder;
+
+impl TextEncoder {
+    pub fn new() -> TextEncoder {
+        TextEncoder
+    }
+}
+
+impl Encoder for TextEncoder {
+    fn encode(&self, metric_familys: &[MetricFamily], writer: &mut Write) -> Result<usize> {
+        let mut written = 0;
+
+        for mf in metric_familys {
+            if mf.get_metric().is_empty() {
+                return Err(Error::Msg("MetricFamily has no metrics".to_owned()));
+            }
+
+            let name = mf.get_name();
+            if name.is_empty() {
+                return Err(Error::Msg("MetricFamily has no name".to_owned()));
+            }
+
+            let help = mf.get_help();
+            if !help.is_empty() {
+                written += try!(
+                writer.write(format!("# HELP {} {}\n", name, escape_string(help, false))
+                        .as_bytes()));
+            }
+
+            let metric_type = mf.get_field_type();
+            let lowercase_type = format!("{:?}", metric_type).to_lowercase();
+            written +=
+                try!(writer.write(format!("# TYPE {} {}\n", name, lowercase_type).as_bytes()));
+
+            for m in mf.get_metric() {
+                match metric_type {
+                    MetricType::COUNTER => {
+                        written += try!(write_sample(name,
+                                                     m,
+                                                     "",
+                                                     "",
+                                                     m.get_counter().get_value(),
+                                                     writer));
+                    }
+                    MetricType::GAUGE | MetricType::SUMMARY | MetricType::HISTOGRAM |
+                    MetricType::UNTYPED => unimplemented!(),
+                }
+            }
+        }
+
+        Ok(written)
+    }
+
+    fn format_type(&self) -> &str {
+        TEXT_FORMAT
+    }
+}
+
+/// `write_sample` writes a single sample in text format to `writer`, given the
+/// metric name, the metric proto message itself, optionally an additional label
+/// name and value (use empty strings if not required), and the value.
+/// The function returns the number of bytes written and any error encountered.
+fn write_sample(name: &str,
+                mc: &proto::Metric,
+                additional_label_name: &str,
+                additional_label_value: &str,
+                value: f64,
+                writer: &mut Write)
+                -> Result<usize> {
+    let mut written = 0;
+
+    written += try!(writer.write(name.as_bytes()));
```

Comment:
I think `write` is ok, `write_sample` is a private function, we guarantee that `name` is not an empty string.